packages/loopring_v3/contracts/iface/ExchangeData.sol

```diff
@@ -72,15 +58,23 @@ library ExchangeData
         uint8 previousMakerFeeBips;
     }
 
+    // General auxiliary data for each conditional transaction
+    struct AuxiliaryData
+    {
+        uint txIndex;
+        bytes data;
+    }
+
     // This is the (virtual) block an operator needs to submit onchain to maintain the
     // per-exchange (virtual) blockchain.
     struct Block
     {
-        ExchangeData.BlockType blockType;
+        uint8                  blockType;
         uint16                 blockSize;
         uint8                  blockVersion;
         bytes                  data;
         uint256[8]             proof;
+        bool                   storeDataHashOnchain;
```

Comment:
The data is only useful if we ever implement those payment receipts, otherwise it's completely useless. So the operator can decide if he needs it to prove if a tx was executed or not. Which I guess will be mostly be not.

> if storeDataHashOnchain is false, we push bytes(0) instead of the data hash, ending up with a few reduced gas cost. I'm not sure this gas savings is worth this extra logic :) Maybe we should just push the data onchain regardless.

You'd actually be surprised as this saves around 20% on the fixed overhead gas costs for each block submission (20,000 gas to store this data onchain (storage is expensive!), it "only" costs 50,000 gas to verify the proof for the block batched). So the smaller this fixed overhead gas cost is, the smaller the blocks can be to still be efficient, which is super important if you check the table of the efficiency for different block sizes in the design doc. So normally I would agree with you for stuff like this, but the two things I want to make as efficient as possible is deposits (because it's paid by users) and block submission (the smaller the blocks, the smaller the trusted setup/prover hardware requirements/... all that stuff).

Reply:
Is there a way to automatically calculate if there are any Transfers inside the block. Instead of asking the operator to do choose the value. I worry the operator will make mistakes. But such an approach will end up most blocks using a `true` value. Can we say the value of `storeDataHashOnchain` is more of a "business decision" instead of a technical decision? If so, it means the transfer receipts are not guaranteed to be useful by the protocol.

Reply:
> Can we say the value of storeDataHashOnchain is more of a "business decision" instead of a technical decision? If so, it means the transfer receipts are not guaranteed to be useful by the protocol.

It's a business decision and receipts are **always** guaranteed to work from the user's perspective. It's the operator that promises to provide a receipt for doing a transfer or not. He needs to prove it, so he's the one that decides if he needs it onchain or not. That is the only purpose of storing that value onchain. The operator **may** choose to provide a user a receipt that he he will do the transfer in a block. If he does so, he needs to make sure he can proof that onchain. In that case, he needs to enable `storeDataHashOnchain`. If he fails to do so, the receipt still works from the user's perspective, but the operator will not be able to proof he did the transfer in the block onchain, so the risk is for the operator.

>  I worry the operator will make mistakes.

The logic is extremely simple for the operator: Did I provide a receipt for a transaction in this block? If yes, enable it, otherwise disable it. That's it. We don't know this on the protocol level because it's something that's build on top of it (which is good).